# Project: Wrangling and Analyze Data

## Data Gathering
In the cell below, gather **all** three pieces of data for this project and load them in the notebook. **Note:** the methods required to gather each data are different.
1. Directly download the WeRateDogs Twitter archive data (twitter_archive_enhanced.csv)

In [1]:
# Import required packages and libraries
import pandas as pd
import numpy as np

# Use `pd.read_csv' to read in `twitter-archive-enhanced.csv' as a dataframe
ratings = pd.read_csv('twitter-archive-enhanced.csv')

# View dataframe using `.head()` to check success
ratings.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

2. Use the Requests library to download the tweet image prediction (image_predictions.tsv)

In [2]:
# Define URL for programmatically accessing `image_predictions.tsv`
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'

# Import requests library and access url using .get() 
import requests
response = requests.get(url)

# Save to file
with open ('image_predictions.tsv', mode='wb') as file:
    file.write(response.content)

# Read in as a dataframe using `pd.read_csv' with argument sep="\t"
images = pd.read_csv('image_predictions.tsv', sep="\t")

# View dataframe using `.head()` to check success
images.head()
    

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

3. Use the Tweepy library to query additional data via the Twitter API (tweet_json.txt)

In [8]:
# Import tweepy library and json library
import tweepy
import json

# Save key and secret variables
consumer_key = 'CONFIDENTIAL'
consumer_secret = 'CONFIDENTIAL'
access_token = 'CONFIDENTIAL'
access_secret = 'CONFIDENTIAL'

#Set API access and tweepy parameters
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)
api = tweepy.API(auth, wait_on_rate_limit=True)


In [9]:
# Define tweet ids to lookup using API
tweet_ids = dog_ratings['tweet_id']

# Lookup tweet data and write to json file
with open('tweet_json.txt', 'w') as outfile:
    for tweet_id in tweet_ids:
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            json.dump(tweet._json, outfile)
            outfile.write('\n')        
        except:
            pass

# Open file and select relevant columns        
data = []
with open('tweet_json.txt', encoding='utf-8') as json_file:
    for line in json_file:
        data.append(json.loads(line))

# Convert to dataframe
tweets_all = pd.DataFrame(data)

# View new dataframe
tweets_all.head()



contributors coordinates                      created_at display_text_range  \
0         None        None  Tue Aug 01 16:23:56 +0000 2017            [0, 85]   
1         None        None  Tue Aug 01 00:17:27 +0000 2017           [0, 138]   
2         None        None  Mon Jul 31 00:18:03 +0000 2017           [0, 121]   
3         None        None  Sun Jul 30 15:58:51 +0000 2017            [0, 79]   
4         None        None  Sat Jul 29 16:00:24 +0000 2017           [0, 138]   

                                            entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...   
1  {'hashtags': [], 'symbols': [], 'user_mentions...   
2  {'hashtags': [], 'symbols': [], 'user_mentions...   
3  {'hashtags': [], 'symbols': [], 'user_mentions...   
4  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   

                                   extended_entities  favorite_count  \
0  {'media': [{'id': 892420639486877696, 'id_str'...           34219   
1  {'media': [{'id': 892177413194625024, 'id_str'...           29671   
2  {'media': [{'id': 891815175371796480, 'id_str'...           22343   
3  {'media': [{'id': 891689552724799489, 'id_str'...           37386   
4  {'media': [{'id': 891327551943041024, 'id_str'...           35733   

   favorited                                          full_text   geo  \
0      False  This is Phineas. He's a mystical boy. Only eve...  None   
1      False  This is Tilly. She's just checking pup on you....  None   
2      False  This is Archie. He is a rare Norwegian Pouncin...  None   
3      False  This is Darla. She commenced a snooze mid meal...  None   
4      False  This is Franklin. He would like you to stop ca...  None   

                         ...                          quoted_status  \
0                        ...                                    NaN   
1                        ...                                    NaN   
2                        ...                                    NaN   
3                        ...                                    NaN   
4                        ...                                    NaN   

  quoted_status_id quoted_status_id_str  quoted_status_permalink  \
0              NaN                  NaN                      NaN   
1              NaN                  NaN                      NaN   
2              NaN                  NaN                      NaN   
3              NaN                  NaN                      NaN   
4              NaN                  NaN                      NaN   

  retweet_count  retweeted retweeted_status  \
0          7099      False              NaN   
1          5350      False              NaN   
2          3520      False              NaN   
3          7302      False              NaN   
4          7842      False              NaN   

                                              source truncated  \
0  <a href="http://twitter.com/download/iphone" r...     False   
1  <a href="http://twitter.com/download/iphone" r...     False   
2  <a href="http://twitter.com/download/iphone" r...     False   
3  <a href="http://twitter.com/download/iphone" r...     False   
4  <a href="http://twitter.com/download/iphone" r...     False   

                                                user  
0  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2  {'id': 4196983835, 'id_str': '4196983835', 'na...  
3  {'id': 4196983835, 'id_str': '4196983835', 'na...  
4  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[5 rows x 32 columns]

In [10]:
# Choose relevant columns from `tweets_all` dataframe to create `tweets` table
retweets = tweets_all[['id', 'retweet_count', 'favorite_count']]

# Export to csv as backup
retweets.to_csv('retweets.csv', index=False)


In [3]:
# Read in tweets.csv as a dataframe
retweets = pd.read_csv('retweets.csv')

# View tweets dataframe to verify success
retweets.head()

id  retweet_count  favorite_count
0  892420643555336193           7099           34219
1  892177421306343426           5350           29671
2  891815181378084864           3520           22343
3  891689557279858688           7302           37386
4  891327558926688256           7842           35733

## Assessing Data
In this section, detect and document at least **eight (8) quality issues and two (2) tidiness issue**. You must use **both** visual assessment
programmatic assessement to assess the data.

**Note:** pay attention to the following key points when you access the data.

* You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.
* Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.
* The fact that the rating numerators are greater than the denominators does not need to be cleaned. This [unique rating system](http://knowyourmeme.com/memes/theyre-good-dogs-brent) is a big part of the popularity of WeRateDogs.
* You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.



In [4]:
# Expand column width so that entire entries can be viewed
pd.set_option('display.max_colwidth', 1)
# Visually assess `ratings` table
ratings.head(25)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0   892420643555336193 NaN                    NaN                    
1   892177421306343426 NaN                    NaN                    
2   891815181378084864 NaN                    NaN                    
3   891689557279858688 NaN                    NaN                    
4   891327558926688256 NaN                    NaN                    
5   891087950875897856 NaN                    NaN                    
6   890971913173991426 NaN                    NaN                    
7   890729181411237888 NaN                    NaN                    
8   890609185150312448 NaN                    NaN                    
9   890240255349198849 NaN                    NaN                    
10  890006608113172480 NaN                    NaN                    
11  889880896479866881 NaN                    NaN                    
12  889665388333682689 NaN                    NaN                    
13  889638837579907072 NaN                    NaN                    
14  889531135344209921 NaN                    NaN                    
15  889278841981685760 NaN                    NaN                    
16  888917238123831296 NaN                    NaN                    
17  888804989199671297 NaN                    NaN                    
18  888554962724278272 NaN                    NaN                    
19  888202515573088257 NaN                    NaN                    
20  888078434458587136 NaN                    NaN                    
21  887705289381826560 NaN                    NaN                    
22  887517139158093824 NaN                    NaN                    
23  887473957103951883 NaN                    NaN                    
24  887343217045368832 NaN                    NaN                    

                    timestamp  \
0   2017-08-01 16:23:56 +0000   
1   2017-08-01 00:17:27 +0000   
2   2017-07-31 00:18:03 +0000   
3   2017-07-30 15:58:51 +0000   
4   2017-07-29 16:00:24 +0000   
5   2017-07-29 00:08:17 +0000   
6   2017-07-28 16:27:12 +0000   
7   2017-07-28 00:22:40 +0000   
8   2017-07-27 16:25:51 +0000   
9   2017-07-26 15:59:51 +0000   
10  2017-07-26 00:31:25 +0000   
11  2017-07-25 16:11:53 +0000   
12  2017-07-25 01:55:32 +0000   
13  2017-07-25 00:10:02 +0000   
14  2017-07-24 17:02:04 +0000   
15  2017-07-24 00:19:32 +0000   
16  2017-07-23 00:22:39 +0000   
17  2017-07-22 16:56:37 +0000   
18  2017-07-22 00:23:06 +0000   
19  2017-07-21 01:02:36 +0000   
20  2017-07-20 16:49:33 +0000   
21  2017-07-19 16:06:48 +0000   
22  2017-07-19 03:39:09 +0000   
23  2017-07-19 00:47:34 +0000   
24  2017-07-18 16:08:03 +0000   

                                                                                source  \
0   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
3   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
4   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
5   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
6   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
7   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
8   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
9   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
10  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
11  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
12  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
13  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter fo

In [5]:
# Programmatically assess `ratings` table
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [6]:
ratings.query('rating_denominator > 10')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
342   832088576586297345  8.320875e+17           3.058208e+07          
433   820690176645140481 NaN                    NaN                    
784   775096608509886464 NaN                    NaN                    
902   758467244762497024 NaN                    NaN                    
1068  740373189193256964 NaN                    NaN                    
1120  731156023742988288 NaN                    NaN                    
1165  722974582966214656 NaN                    NaN                    
1202  716439118184652801 NaN                    NaN                    
1228  713900603437621249 NaN                    NaN                    
1254  710658690886586372 NaN                    NaN                    
1274  709198395643068416 NaN                    NaN                    
1351  704054845121142784 NaN                    NaN                    
1433  697463031882764288 NaN                    NaN                    
1598  686035780142297088  6.860340e+17           4.196984e+09          
1634  684225744407494656  6.842229e+17           4.196984e+09          
1635  684222868335505415 NaN                    NaN                    
1662  682962037429899265 NaN                    NaN                    
1663  682808988178739200  6.827884e+17           4.196984e+09          
1779  677716515794329600 NaN                    NaN                    
1843  675853064436391936 NaN                    NaN                    

                      timestamp  \
342   2017-02-16 04:45:50 +0000   
433   2017-01-15 17:52:40 +0000   
784   2016-09-11 22:20:06 +0000   
902   2016-07-28 01:00:57 +0000   
1068  2016-06-08 02:41:38 +0000   
1120  2016-05-13 16:15:54 +0000   
1165  2016-04-21 02:25:47 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1598  2016-01-10 04:04:10 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1662  2016-01-01 16:30:13 +0000   
1663  2016-01-01 06:22:03 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   

                                                                                  source  \
342   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
433   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
784   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
902   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1068  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1120  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1165  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1202  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1228  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1254  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1274  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1351  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1433  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1598  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1634  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1635  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1662  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1663  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1779  <a href="http

In [7]:
ratings['rating_numerator'].value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7       55 
14      54 
5       37 
6       32 
3       19 
4       17 
1       9  
2       9  
420     2  
0       2  
15      2  
75      2  
80      1  
20      1  
24      1  
26      1  
44      1  
50      1  
60      1  
165     1  
84      1  
88      1  
144     1  
182     1  
143     1  
666     1  
960     1  
1776    1  
17      1  
27      1  
45      1  
99      1  
121     1  
204     1  
Name: rating_numerator, dtype: int64

In [8]:
ratings.query('rating_numerator ==5')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
45    883482846933004288 NaN                    NaN                    
730   781661882474196992 NaN                    NaN                    
956   751583847268179968 NaN                    NaN                    
1399  699691744225525762 NaN                    NaN                    
1461  694925794720792577 NaN                    NaN                    
1508  691483041324204033 NaN                    NaN                    
1583  687102708889812993 NaN                    NaN                    
1618  684969860808454144  6.849598e+17           4.196984e+09          
1619  684959798585110529 NaN                    NaN                    
1624  684880619965411328 NaN                    NaN                    
1645  683849932751646720 NaN                    NaN                    
1680  682003177596559360 NaN                    NaN                    
1689  681340665377193984  6.813394e+17           4.196984e+09          
1727  679877062409191424 NaN                    NaN                    
1796  677301033169788928 NaN                    NaN                    
1808  676897532954456065 NaN                    NaN                    
1820  676588346097852417 NaN                    NaN                    
1861  675483430902214656 NaN                    NaN                    
1874  675135153782571009 NaN                    NaN                    
1901  674646392044941312 NaN                    NaN                    
1904  674632714662858753 NaN                    NaN                    
1925  674063288070742018 NaN                    NaN                    
1979  672980819271634944 NaN                    NaN                    
2013  672231046314901505 NaN                    NaN                    
2026  671879137494245376 NaN                    NaN                    
2063  671159727754231808 NaN                    NaN                    
2092  670782429121134593 NaN                    NaN                    
2109  670449342516494336 NaN                    NaN                    
2134  670069087419133954 NaN                    NaN                    
2139  670037189829525505 NaN                    NaN                    
2153  669661792646373376 NaN                    NaN                    
2181  668994913074286592 NaN                    NaN                    
2206  668631377374486528 NaN                    NaN                    
2242  667911425562669056 NaN                    NaN                    
2260  667550882905632768 NaN                    NaN                    
2312  666776908487630848 NaN                    NaN                    
2351  666049248165822465 NaN                    NaN                    

                      timestamp  \
45    2017-07-08 00:28:19 +0000   
730   2016-09-30 01:08:10 +0000   
956   2016-07-09 01:08:47 +0000   
1399  2016-02-16 20:28:06 +0000   
1461  2016-02-03 16:49:55 +0000   
1508  2016-01-25 04:49:38 +0000   
1583  2016-01-13 02:43:46 +0000   
1618  2016-01-07 05:28:35 +0000   
1619  2016-01-07 04:48:36 +0000   
1624  2016-01-06 23:33:58 +0000   
1645  2016-01-04 03:18:23 +0000   
1680  2015-12-30 01:00:03 +0000   
1689  2015-12-28 05:07:27 +0000   
1727  2015-12-24 04:11:37 +0000   
1796  2015-12-17 01:35:24 +0000   
1808  2015-12-15 22:52:02 +0000   
1820  2015-12-15 02:23:26 +0000   
1861  2015-12-12 01:12:54 +0000   
1874  2015-12-11 02:08:58 +0000   
1901  2015-12-09 17:46:48 +0000   
1904  2015-12-09 16:52:27 +0000   
1925  2015-12-08 03:09:46 +0000   
1979  2015-12-05 03:28:25 +0000   
2013  2015-12-03 01:49:05 +0000   
2026  2015-12-02 02:30:43 +0000   
2063  2015-11-30 02:52:03 +0000   
2092  2015-11-29 01:52:48 +0000   
2109  2015-11-28 03:49:14 +0000   
2134  2015-11-27 02:38:14 +0000   
2139  2015-11-27 00:31:29 +0000   
2153  2015-11-25 23:39:47 +0000   
2181  2015-11-24 03:29:51 +0000   
2206  2015-11-23 03:25:17 +0000   
2242  2015-11-21 03:44:27 +0000   
2260  2015-11-20 03:51:47 +0000   
2312  2015-11-18 00

In [9]:
ratings.query('tweet_id == 666020888022790149')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2355  666020888022790149 NaN                    NaN                    

                      timestamp  \
2355  2015-11-15 22:32:08 +0000   

                                                                                  source  \
2355  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                     text  \
2355  Here we have a Japanese Irish Setter. Lost eye in Vietnam (?). Big fan of relaxing on stair. 8/10 would pet https://t.co/BLDqew2Ijj   

      retweeted_status_id  retweeted_status_user_id  \
2355 NaN                  NaN                         

     retweeted_status_timestamp  \
2355  NaN                         

                                                        expanded_urls  \
2355  https://twitter.com/dog_rates/status/666020888022790149/photo/1   

      rating_numerator  rating_denominator  name doggo floofer pupper puppo  
2355  8                 10                  None  None  None    None   None

In [10]:
images.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [11]:
images.head(25)

tweet_id                                          jpg_url  \
0   666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1   666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2   666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3   666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4   666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5   666050758794694657  https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6   666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7   666055525042405380  https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8   666057090499244032  https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9   666058600524156928  https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10  666063827256086533  https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11  666071193221509120  https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12  666073100786774016  https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13  666082916733198337  https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14  666094000022159362  https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15  666099513787052032  https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16  666102155909144576  https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17  666104133288665088  https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18  666268910803644416  https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19  666273097616637952  https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20  666287406224695296  https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21  666293911632134144  https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22  666337882303524864  https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23  666345417576210432  https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24  666353288456101888  https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   

    img_num                      p1   p1_conf  p1_dog  \
0   1        Welsh_springer_spaniel  0.465074  True     
1   1        redbone                 0.506826  True     
2   1        German_shepherd         0.596461  True     
3   1        Rhodesian_ridgeback     0.408143  True     
4   1        miniature_pinscher      0.560311  True     
5   1        Bernese_mountain_dog    0.651137  True     
6   1        box_turtle              0.933012  False    
7   1        chow                    0.692517  True     
8   1        shopping_cart           0.962465  False    
9   1        miniature_poodle        0.201493  True     
10  1        golden_retriever        0.775930  True     
11  1        Gordon_setter           0.503672  True     
12  1        Walker_hound            0.260857  True     
13  1        pug                     0.489814  True     
14  1        bloodhound              0.195217  True     
15  1        Lhasa                   0.582330  True     
16  1        English_setter          0.298617  True     
17  1        hen                     0.965932  False    
18  1        desktop_computer        0.086502  False    
19  1        Italian_greyhound       0.176053  True     
20  1        Maltese_dog             0.857531  True     
21  1        three-toed_sloth        0.914671  False    
22  1        ox                      0.416669  False    
23  1        golden_retriever        0.858744  True     
24  1        malamute                0.336874  True     

                          p2   p2_conf  p2_dog                           p3  \
0   collie                    0.156665  True    Shetland_sheepdog             
1   miniature_pinscher        0.074192  True    Rhodesian_ridgeback           
2   malinois                  0.138584  True    bloodhound                    
3   redbone                   0.360687  True    miniature_pinscher            
4   Rottweiler                0.243682  True    Doberman                      
5   English_springer          0.263788  True    Greater_Swiss_Mountain_dog    
6   mud_turtle             

In [12]:
retweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2327 entries, 0 to 2326
Data columns (total 3 columns):
id                2327 non-null int64
retweet_count     2327 non-null int64
favorite_count    2327 non-null int64
dtypes: int64(3)
memory usage: 54.6 KB


### Quality issues
1. There are 181 retweets and 78 replies in the `ratings` table i.e. there are non-null entries in the 'in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id' and 'retweeted_status_timestamp' columns. These are not original dog rating tweets. 

2. There are 59 missing values in the 'expanded_urls' column for the `ratings` table. Some entries in this column have multiple urls that are duplicates of each other.

3. 'rating_numerator' and 'rating_denominator' columns in the `dog_ratings` table need to to allow for decimals such as entry 45. 'id' and 'tweet_id' columns do not require numerical operations to be carried out. 'timestamp' and 'retweeted_status_timestamp' are not datetime objects.

4. Value for 'name' is incorrect for some entries in the `ratings` table e.g. 'name' is recorded as *"such"* for entry in row 22.

5. Null values are recorded as *"None"* instead of NaN for 'name', 'doggo', 'floofer', 'pupper' and 'puppo' columns in `ratings` table.

6. Values for 'rating_numerator' and 'rating_denominator' are incorrect for some entries in the `ratings` table e.g. recorded as *"9"* and *"11"* for entry in row 1068 when they should be "14" and "10".

7. Entries in 'p1', 'p2' and 'p3' columns in the `images` table are not consistent in case.

8. 'id' column name in `retweets` table does not match the 'tweet_id' column in the `dog_ratings` and `images` tables

   

### Tidiness issues
1. Retweets and favorites are handled in the retweets table. 'in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id' and 'retweeted_status_timestamp' are not necessary in the ratings table as there are only original tweets in this table.

2. Dog 'stage' variable is spread across 4 columns with values 'doggo', 'floofer', 'pupper' and 'puppo' in the `ratings` table. 


## Cleaning Data
In this section, clean **all** of the issues you documented while assessing. 

**Note:** Make a copy of the original data before cleaning. Cleaning includes merging individual pieces of data according to the rules of [tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html). The result should be a high-quality and tidy master pandas DataFrame (or DataFrames, if appropriate).

In [13]:
# Make copies of original pieces of data
ratings_clean = ratings.copy()
images_clean = images.copy()
retweets_clean = retweets.copy()

### Quality Issue #1 :
* There are 181 retweets and 78 replies in the `ratings` table i.e. there are non-null entries in the 'in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id' and 'retweeted_status_timestamp' columns. These are not original dog rating tweets. 

#### Define:
* Filter data to entries which have a null value in the 'in_reply_to_status_id' or 'retweeted_status_id' columns.

#### Code:

In [14]:
# Filter data to entries which have a null value in the 'in_reply_to_status_id' column
ratings_clean = ratings_clean[ratings_clean['retweeted_status_id'].isnull() == True]
# Filter data to entries which have a null value in the 'retweeted_status_id' column
ratings_clean = ratings_clean[ratings_clean['in_reply_to_status_id'].isnull() == True]


#### Test:

In [15]:
# Verify that all entries in the 'in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id',
#'retweeted_status_user_id' and 'retweeted_status_timestamp' columns are null values
ratings_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2097 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2097 non-null object
source                        2097 non-null object
text                          2097 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2094 non-null object
rating_numerator              2097 non-null int64
rating_denominator            2097 non-null int64
name                          2097 non-null object
doggo                         2097 non-null object
floofer                       2097 non-null object
pupper                        2097 non-null object
puppo                         2097 non-null object
dtypes: float64(4), int64(3), object(10)

### Quality Issue #2: 
* There are 59 missing values in the 'expanded_urls' column for the `ratings` table. Some entries in this column have multiple urls that are duplicates of each other.

#### Define:
* Use `str.extract` to extract the url information from the 'text' column and save in a new 'url' column. Drop the 'expanded_urls' column.

#### Code:

In [16]:
# Use str.extract to extract the url information from the 'text' column and save in a new 'url' column.
ratings_clean['url'] = ratings_clean['text'].str.extract(r"([https]+://[\w(\.|/)]+)")
# # Check for null values in new column
ratings_clean[ratings_clean['url'].isnull() == True]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
375   828361771580813312 NaN                    NaN                    
707   785515384317313025 NaN                    NaN                    
1445  696518437233913856 NaN                    NaN                    

                      timestamp  \
375   2017-02-05 21:56:51 +0000   
707   2016-10-10 16:20:36 +0000   
1445  2016-02-08 02:18:30 +0000   

                                                                                  source  \
375   <a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                   
707   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
1445  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                           text  \
375   Beebop and Doobert should start a band 12/10 would listen   
707   Today, 10/10, should be National Dog Rates Day              
1445  Oh my god 10/10 for every little hot dog pupper             

      retweeted_status_id  retweeted_status_user_id  \
375  NaN                  NaN                         
707  NaN                  NaN                         
1445 NaN                  NaN                         

     retweeted_status_timestamp expanded_urls  rating_numerator  \
375   NaN                        NaN           12                 
707   NaN                        NaN           10                 
1445  NaN                        NaN           10                 

      rating_denominator  name doggo floofer  pupper puppo  url  
375   10                  None  None  None    None    None  NaN  
707   10                  None  None  None    None    None  NaN  
1445  10                  None  None  None    pupper  None  NaN

In [17]:
# Manually drop these three rows as the 'text' field does not contain a URL and the 'expanded_URLs' field was also empty
ratings_clean.drop(labels=[375, 707, 1445], axis=0, inplace=True)

# Drop the 'expanded_urls' column
ratings_clean.drop(columns='expanded_urls', inplace=True)


#### Test:

In [18]:
# Filter data by null values in the 'url' column - this should be empty
ratings_clean[ratings_clean['url'].isnull() == True]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo, url]
Index: []

In [19]:
# Ensure 'expanded_urls' column has been removed
ratings_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2094 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2094 non-null int64
in_reply_to_status_id         0 non-null float64
in_reply_to_user_id           0 non-null float64
timestamp                     2094 non-null object
source                        2094 non-null object
text                          2094 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
rating_numerator              2094 non-null int64
rating_denominator            2094 non-null int64
name                          2094 non-null object
doggo                         2094 non-null object
floofer                       2094 non-null object
pupper                        2094 non-null object
puppo                         2094 non-null object
url                           2094 non-null object
dtypes: float64(4), int64(3), object(10)

### Tidiness Issue #1: 
* Retweets and favorites are handled in the retweets table. 'in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id' and 'retweeted_status_timestamp' columns are not necessary in the `ratings` table as there are only original tweets in this table.

#### Define:
* Drop the 'in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id' and 'retweeted_status_timestamp' columns from the `ratings` table.

#### Code:

In [20]:
# Drop the 'in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id' and 
#'retweeted_status_timestamp' columns
ratings_clean.drop(columns=['in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], inplace=True)

#### Test:

In [21]:
# Verify that columns have been removed
ratings_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2094 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2094 non-null int64
timestamp             2094 non-null object
source                2094 non-null object
text                  2094 non-null object
rating_numerator      2094 non-null int64
rating_denominator    2094 non-null int64
name                  2094 non-null object
doggo                 2094 non-null object
floofer               2094 non-null object
pupper                2094 non-null object
puppo                 2094 non-null object
url                   2094 non-null object
dtypes: int64(3), object(9)
memory usage: 212.7+ KB


### Quality Issue #3:
*'rating_numerator' and 'rating_denominator' columns in the dog_ratings table need to to allow for decimals such as entry 45. 'id' and 'tweet_id' columns do not require numerical operations to be carried out. 'timestamp' and 'retweeted_status_timestamp' are not datetime objects.

#### Define:
* Convert 'rating_numerator' and 'rating_denominator' columns to float datatype
* Convert 'id' and 'tweet_id' columns to string datatype
* Convert 'timestamp' column to datetime datatype
* 'retweeted_status_timestamp' has been removed above


#### Code:

In [22]:
# Convert 'rating_numerator' and 'rating_denominator' columns to float datatype
ratings_clean[['rating_numerator', 'rating_denominator']] = ratings_clean[['rating_numerator', 'rating_denominator']].astype(float)

# Convert 'id' and 'tweet_id' columns to string datatype
ratings_clean['tweet_id'] = ratings_clean['tweet_id'].astype(str)
images_clean['tweet_id'] = images_clean['tweet_id'].astype(str)
retweets_clean['id'] = retweets_clean['id'].astype(str)

# Convert 'timestamp' column to datetime datatype
ratings_clean['timestamp'] = pd.to_datetime(ratings_clean['timestamp'])


#### Test:

In [23]:
# Check datatypes for each relevant column
ratings_clean.info()
images_clean.info()
retweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2094 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2094 non-null object
timestamp             2094 non-null datetime64[ns]
source                2094 non-null object
text                  2094 non-null object
rating_numerator      2094 non-null float64
rating_denominator    2094 non-null float64
name                  2094 non-null object
doggo                 2094 non-null object
floofer               2094 non-null object
pupper                2094 non-null object
puppo                 2094 non-null object
url                   2094 non-null object
dtypes: datetime64[ns](1), float64(2), object(9)
memory usage: 212.7+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null object
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-n

### Quality Issue #4 :
* Value for 'name' is incorrect for some entries in the `ratings` table e.g. 'name' is recorded as *"such"* for entry in row 22.

#### Define:
* Use `str.extract` to extract regular expressions that match common patterns from 'name' column. Extract the value for name and remove the "." from each entry.


#### Code

In [24]:
# Use `str.extract` to extract regular expressions that match common patterns from 'name' column.
ratings_clean['name'] = ratings_clean['text'].str.extract(r"(.*is\s[A-Z][a-z]*.|Meet\s[A-Z][a-z]*.|Say\shello\sto\s[A-Z][a-z]*.)")
# Extract the value for name
ratings_clean['name'] = ratings_clean['name'].str.extract(r"([A-Z][a-z]*\.)")
# Remove the "." from each entry.
ratings_clean['name'] = ratings_clean['name'].str.replace('.', '')


#### Test

In [25]:
# View first 25 entries in 'name' column, entry 22 should be Nan 
ratings_clean['name'].head(25) 

0     Phineas 
1     Tilly   
2     Archie  
3     Darla   
4     Franklin
5     NaN     
6     Jax     
7     NaN     
8     Zoey    
9     Cassie  
10    Koda    
11    Bruno   
12    NaN     
13    Ted     
14    Stuart  
15    Oliver  
16    Jim     
17    Zeke    
18    Ralphus 
20    Gerald  
21    Jeffrey 
22    NaN     
23    Canela  
24    NaN     
25    NaN     
Name: name, dtype: object

In [26]:
# Check that all "None" entries have been removed 
ratings_clean.query('name == "None"')

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo, url]
Index: []

### Quality Issue #5:
* Null values are recorded as *"None"* instead of NaN for 'name', 'doggo', 'floofer', 'pupper' and 'puppo' columns in `ratings` table.

#### Define
* 'name' column has been addressed above
* Use `str.extract` to extract regular expressions that match 'doggo', 'floofer', 'pupper' and 'puppo' for each column 

#### Code

In [27]:
# Use `str.extract` to extract regular expressions that match 'doggo', 'floofer', 'pupper' and 'puppo' for each column 
ratings_clean['doggo'] = ratings_clean['doggo'].str.extract(r"(\bdoggo\b)")
ratings_clean['floofer'] = ratings_clean['floofer'].str.extract(r"(\bfloofer\b)")
ratings_clean['pupper'] = ratings_clean['pupper'].str.extract(r"(\bpupper\b)")
ratings_clean['puppo'] = ratings_clean['puppo'].str.extract(r"(\bpuppo\b)")


#### Test

In [28]:
# Check that all "None" entries have been removed 
ratings_clean.query('doggo == "None"' or 'floofer == "None"' or 'pupper == "None"' or 'puppo == "None"')

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo, url]
Index: []

### Tidiness Issue #2:
* Dog 'stage' variable is spread across 4 columns with values 'doggo', 'floofer', 'pupper' and 'puppo' in the ratings table.

#### Define:
* Concatenate the data for 'doggo', 'floofer', 'pupper' and 'puppo' columns into a single column
* Drop the 'doggo', 'floofer', 'pupper' and 'puppo' columns

#### Code:

In [29]:
# Concatenate the data for 'doggo', 'floofer', 'pupper' and 'puppo' columns into a single column
strings = ratings_clean[['floofer', 'pupper', 'puppo']]
ratings_clean['stage']= ratings_clean['doggo'].str.cat(strings, sep ="", na_rep="")

# Check output
ratings_clean['stage'].value_counts()


                1759
pupper          220 
doggo           72  
puppo           23  
floofer         9   
doggopupper     9   
doggofloofer    1   
doggopuppo      1   
Name: stage, dtype: int64

In [30]:
# Replace empty strings with NaN
ratings_clean['stage'] = ratings_clean['stage'].replace(r'^\s*$', np.NaN, regex=True)

# Separate 'doggopupper', 'doggofloofer', 'doggopuppo' strings into two values
ratings_clean[ratings_clean['stage'] == 'doggopupper'] = 'doggo, pupper'
ratings_clean[ratings_clean['stage'] == 'doggofloofer'] = 'doggo, floofer'
ratings_clean[ratings_clean['stage'] == 'doggopuppo'] = 'doggo, puppo'

# Drop the 'doggo', 'floofer', 'pupper' and 'puppo' columns
ratings_clean.drop(columns=['doggo', 'floofer', 'pupper','puppo'], inplace=True)


#### Test:

In [31]:
# Check value counts for new 'stage' column
ratings_clean['stage'].value_counts()

pupper            220
doggo             72 
puppo             23 
doggo, pupper     9  
floofer           9  
doggo, floofer    1  
doggo, puppo      1  
Name: stage, dtype: int64

In [32]:
# Ensure columns have been removed
ratings_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2094 entries, 0 to 2355
Data columns (total 9 columns):
tweet_id              2094 non-null object
timestamp             2094 non-null object
source                2094 non-null object
text                  2094 non-null object
rating_numerator      2094 non-null object
rating_denominator    2094 non-null object
name                  1310 non-null object
url                   2094 non-null object
stage                 335 non-null object
dtypes: object(9)
memory usage: 163.6+ KB


### Issue:
* Values for 'rating_numerator' and 'rating_denominator' are incorrect for some entries in the `ratings` table e.g. recorded as *"9"* and *"11"* for entry in row 1068 when they should be "14" and "10".

#### Define:
* Use `str.extract` to extract the full rating from the 'text' column using a regular expression and store in a 'full_rating' column
* Use str.split to extract and store the numerator and denominator in correct columns

#### Code:

#### Test:

## Storing Data
Save gathered, assessed, and cleaned master dataset to a CSV file named "twitter_archive_master.csv".

## Analyzing and Visualizing Data
In this section, analyze and visualize your wrangled data. You must produce at least **three (3) insights and one (1) visualization.**

### Insights:
1.

2.

3.

### Visualization